File: scripts/build-times.js
#!/usr/bin/env node

import { execSync } from 'node:child_process'
import fs from 'node:fs/promises'
import path from 'node:path'
import { performance } from 'node:perf_hooks'
import process from 'node:process'
import pc from 'picocolors'

async function runBuild(name, directory, command) {
  console.warn(pc.bold(`\n🔨 Building ${name}...`))
  console.warn(pc.dim(`Directory: ${directory}`))
  console.warn(pc.dim(`Command: ${command}`))

  const startTime = performance.now()

  try {
    const result = execSync(command, {
      cwd: path.resolve(`${directory}`),
      env: { ...process.env, NODE_ENV: 'production' },
      encoding: 'utf8',
      timeout: 120000,
    })

    const endTime = performance.now()
    const duration = endTime - startTime

    console.warn(pc.green(`  ✅ ${name} built successfully in ${duration.toFixed(2)}ms`))

    return {
      success: true,
      duration,
      stdout: result,
      stderr: '',
      exitCode: 0,
    }
  }
  catch (error) {
    const endTime = performance.now()
    const duration = endTime - startTime

    console.warn(pc.red(`  ❌ ${name} build failed (exit code ${error.status || -1})`))
    console.warn(pc.red(`  Error: ${error.stderr || error.message}`))

    return {
      success: false,
      duration,
      stdout: error.stdout || '',
      stderr: error.stderr || error.message,
      exitCode: error.status || -1,
    }
  }
}

async function getClientBundleSize(name, directory) {
  let totalSize = 0
  let chunkCount = 0

  if (name === 'Next.js') {
    const staticDir = path.resolve(`${directory}/.next/static`)
    const chunksDir = path.join(staticDir, 'chunks')

    try {
      const chunkFiles = await fs.readdir(chunksDir, { recursive: true })
      for (const file of chunkFiles) {
        if (file.endsWith('.js')) {
          chunkCount++
          const filePath = path.join(chunksDir, file)
          try {
            const stats = await fs.stat(filePath)
            totalSize += stats.size
          }
          catch {
          }
        }
      }
    }
    catch {
    }
  }
  else if (name === 'Rari') {
    const distDir = path.resolve(`${directory}/dist`)
    try {
      const files = await fs.readdir(distDir, { recursive: true })
      for (const file of files) {
        if (file.endsWith('.js') || file.endsWith('.css')) {
          chunkCount++
          const filePath = path.join(distDir, file)
          try {
            const stats = await fs.stat(filePath)
            totalSize += stats.size
          }
          catch {
          }
        }
      }
    }
    catch {
    }
  }

  return { totalSize, chunkCount }
}

async function analyzeBuildOutput(name, result) {
  if (!result.success) {
    return { error: 'Build failed' }
  }

  const output = result.stdout + result.stderr

  const analysis = {
    duration: result.duration,
    bundleSize: null,
    chunkCount: null,
    warnings: 0,
    errors: 0,
  }

  analysis.warnings = (output.match(/warning/gi) || []).length
  analysis.errors = (output.match(/error/gi) || []).length

  const directory = name === 'Next.js' ? 'nextjs-app' : 'rari-app'
  const bundleInfo = await getClientBundleSize(name, directory)

  analysis.chunkCount = bundleInfo.chunkCount
  if (bundleInfo.totalSize > 0) {
    const sizeInKB = bundleInfo.totalSize / 1024
    analysis.bundleSize = `${sizeInKB.toFixed(2)} kB`
  }

  return analysis
}

function displayComparison(rariResult, nextjsResult) {
  console.warn(pc.bold('\n📊 Build Performance Comparison'))

  console.warn('\n⏱️  Build Times:')
  console.warn(`  🦀 Rari:     ${(rariResult.duration / 1000).toFixed(2)}s`)
  console.warn(`  🟢 Next.js:  ${(nextjsResult.duration / 1000).toFixed(2)}s`)

  const timeDiff = ((rariResult.duration - nextjsResult.duration) / nextjsResult.duration) * 100
  if (timeDiff < 0) {
    console.warn(pc.green(`  📈 Rari builds ${Math.abs(timeDiff).toFixed(1)}% faster`))
  }
  else {
    console.warn(pc.red(`  📉 Rari builds ${timeDiff.toFixed(1)}% slower`))
  }

  console.warn('\n📦 Bundle Information:')
  console.warn(`  🦀 Rari:`)
  console.warn(`     Size: ${rariResult.bundleSize || 'Unknown'}`)
  console.warn(`     Chunks: ${rariResult.chunkCount || 'Unknown'}`)
  console.warn(`     Warnings: ${rariResult.warnings}`)
  console.warn(`     Errors: ${rariResult.errors}`)

  console.warn(`  🟢 Next.js:`)
  console.warn(`     Size: ${nextjsResult.bundleSize || 'Unknown'}`)
  console.warn(`     Chunks: ${nextjsResult.chunkCount || 'Unknown'}`)
  console.warn(`     Warnings: ${nextjsResult.warnings}`)
  console.warn(`     Errors: ${nextjsResult.errors}`)
}

async function saveResults(rariResult, nextjsResult) {
  const timestamp = new Date().toISOString()
  const results = {
    timestamp,
    rari: rariResult,
    nextjs: nextjsResult,
  }

  try {
    await fs.mkdir('results', { recursive: true })
  }
  catch {
  }

  await fs.writeFile(
    `./results/buildtimes-${timestamp.slice(0, 10)}.json`,
    JSON.stringify(results, null, 2),
  )

  console.warn(pc.dim(`\n💾 Results saved to ./results/buildtimes-${timestamp.slice(0, 10)}.json`))
}

async function main() {
  console.warn(pc.bold(pc.cyan('🔨 Rari vs Next.js Build Time Comparison')))
  console.warn(pc.dim('This benchmark compares build performance and bundle analysis\n'))

  console.warn(pc.yellow('⚠️  This will run production builds which may take some time'))
  console.warn(pc.dim('Starting build comparison in 3 seconds...'))
  await new Promise(resolve => setTimeout(resolve, 3000))

  const rariResult = await runBuild('Rari', 'rari-app', 'pnpm run build')
  const rariAnalysis = await analyzeBuildOutput('Rari', rariResult)

  const nextjsResult = await runBuild('Next.js', 'nextjs-app', 'pnpm run build')
  const nextjsAnalysis = await analyzeBuildOutput('Next.js', nextjsResult)

  displayComparison(rariAnalysis, nextjsAnalysis)
  await saveResults(rariAnalysis, nextjsAnalysis)

  console.warn(pc.bold(pc.green('\n🎉 Build comparison completed!')))
}

main().catch(console.error)

File: tools/benchmark/src/build_times.rs
mod time_utils;

use anyhow::{Context, Result};
use clap::Parser;
use colored::Colorize;
use serde::{Deserialize, Serialize};
use std::future::Future;
use std::path::{Path, PathBuf};
use std::pin::Pin;
use std::time::{Instant, SystemTime};
use tokio::fs;
use tokio::process::Command;

#[derive(Parser, Debug)]
#[command(name = "build-times")]
#[command(about = "Compare build times between rari and Next.js")]
struct Args {
    #[arg(short, long, default_value = ".")]
    dir: PathBuf,
}

#[derive(Debug, Serialize, Deserialize)]
struct BuildResult {
    success: bool,
    duration_ms: f64,
    bundle_size: Option<String>,
    chunk_count: Option<usize>,
    warnings: usize,
    errors: usize,
}

#[derive(Debug, Serialize, Deserialize)]
struct BenchmarkResults {
    timestamp: String,
    rari: BuildResult,
    nextjs: BuildResult,
}

async fn run_build(name: &str, directory: &Path, command: &str) -> Result<BuildResult> {
    println!("\n{} Building {}...", "🔨".bold(), name.bold());
    println!("  {} {}", "Directory:".dimmed(), directory.display());
    println!("  {} {}", "Command:".dimmed(), command);

    let start = Instant::now();

    let parts: Vec<&str> = command.split_whitespace().collect();
    let (cmd, args) = parts.split_first().context("Empty command")?;

    let output = Command::new(cmd)
        .args(args)
        .current_dir(directory)
        .env("NODE_ENV", "production")
        .output()
        .await
        .context("Failed to execute build command")?;

    let duration = start.elapsed();
    let duration_ms = duration.as_secs_f64() * 1000.0;

    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);
    let combined = format!("{}{}", stdout, stderr);

    let success = output.status.success();

    if success {
        println!(
            "  {} {} built successfully in {:.2}ms",
            "✅".green(),
            name,
            duration_ms
        );
    } else {
        println!(
            "  {} {} build failed (exit code {})",
            "❌".red(),
            name,
            output.status.code().unwrap_or(-1)
        );
        if !stderr.is_empty() {
            println!("  {} {}", "Error:".red(), stderr.trim());
        }
    }

    let deprecation_warnings =
        combined.matches("DeprecationWarning").count() + combined.matches("[DEP").count();

    let warnings = (combined.matches("warning").count()
        + combined.matches("Warning").count()
        + combined.matches("WARNING").count())
    .saturating_sub(deprecation_warnings);

    let errors = combined.matches("error").count()
        + combined.matches("Error").count()
        + combined.matches("ERROR").count();

    let (bundle_size, chunk_count) = if success {
        get_bundle_info(name, directory).await?
    } else {
        (None, None)
    };

    Ok(BuildResult {
        success,
        duration_ms,
        bundle_size,
        chunk_count,
        warnings,
        errors,
    })
}

async fn get_bundle_info(name: &str, directory: &Path) -> Result<(Option<String>, Option<usize>)> {
    let (dist_dir, extensions): (PathBuf, Vec<&str>) = match name {
        "Next.js" => (directory.join(".next/static/chunks"), vec!["js", "css"]),
        "rari" => (directory.join("dist/assets"), vec!["js", "css"]),
        _ => return Ok((None, None)),
    };

    if !dist_dir.exists() {
        return Ok((None, None));
    }

    let mut total_size = 0u64;
    let mut chunk_count = 0usize;

    scan_directory(&dist_dir, &extensions, &mut total_size, &mut chunk_count).await?;

    let bundle_size = if total_size > 0 {
        Some(format!("{:.2} kB", total_size as f64 / 1024.0))
    } else {
        None
    };

    Ok((bundle_size, Some(chunk_count)))
}

fn scan_directory<'a>(
    dir: &'a Path,
    extensions: &'a [&'a str],
    total_size: &'a mut u64,
    chunk_count: &'a mut usize,
) -> Pin<Box<dyn Future<Output = Result<()>> + 'a>> {
    Box::pin(async move {
        let mut entries = fs::read_dir(dir).await?;
        while let Some(entry) = entries.next_entry().await? {
            let path = entry.path();
            if path.is_dir() {
                scan_directory(&path, extensions, total_size, chunk_count).await?;
            } else if path.is_file()
                && let Some(ext) = path.extension()
                && extensions.contains(&ext.to_str().unwrap_or(""))
                && let Ok(metadata) = fs::metadata(&path).await
            {
                *total_size += metadata.len();
                *chunk_count += 1;
            }
        }
        Ok(())
    })
}

fn display_comparison(rari: &BuildResult, nextjs: &BuildResult) {
    println!("\n{}", "📊 Build Performance Comparison".bold());

    println!("\n⏱️  Build Times:");
    println!("  🦀 rari:     {:.2}s", rari.duration_ms / 1000.0);
    println!("  🟢 Next.js:  {:.2}s", nextjs.duration_ms / 1000.0);

    let time_diff = ((rari.duration_ms - nextjs.duration_ms) / nextjs.duration_ms) * 100.0;
    if time_diff < 0.0 {
        println!(
            "  {} rari builds {:.1}% faster",
            "📈".green(),
            time_diff.abs()
        );
    } else {
        println!("  {} rari builds {:.1}% slower", "📉".red(), time_diff);
    }

    println!("\n📦 Client Bundle Information:");
    println!("  🦀 rari:");
    println!(
        "     Size: {}",
        rari.bundle_size.as_deref().unwrap_or("Unknown")
    );
    println!(
        "     Files: {}",
        rari.chunk_count
            .map_or("Unknown".to_string(), |c| c.to_string())
    );
    println!("     Warnings: {}", rari.warnings);
    println!("     Errors: {}", rari.errors);

    println!("  🟢 Next.js:");
    println!(
        "     Size: {}",
        nextjs.bundle_size.as_deref().unwrap_or("Unknown")
    );
    println!(
        "     Files: {}",
        nextjs
            .chunk_count
            .map_or("Unknown".to_string(), |c| c.to_string())
    );
    println!("     Warnings: {}", nextjs.warnings);
    println!("     Errors: {}", nextjs.errors);
}

async fn save_results(results: &BenchmarkResults, base_dir: &Path) -> Result<()> {
    let results_dir = base_dir.join("results");
    fs::create_dir_all(&results_dir).await?;

    let date = time_utils::format_date(SystemTime::now());
    let filename = results_dir.join(format!("buildtimes-{}.json", date));

    let json = format!("{}\n", serde_json::to_string_pretty(results)?);
    fs::write(&filename, json).await?;

    println!(
        "\n{} Results saved to {}",
        "💾".dimmed(),
        filename.display()
    );

    Ok(())
}

#[tokio::main]
async fn main() -> Result<()> {
    let args = Args::parse();

    println!(
        "{}",
        "🔨 rari vs Next.js Build Time Comparison".cyan().bold()
    );
    println!(
        "{}",
        "This benchmark compares build performance and bundle analysis\n".dimmed()
    );

    println!(
        "{}",
        "⚠️  This will run production builds which may take some time".yellow()
    );
    println!("{}", "Starting build comparison in 3 seconds...".dimmed());
    tokio::time::sleep(tokio::time::Duration::from_secs(3)).await;

    let rari_dir = args.dir.join("rari-app");
    let nextjs_dir = args.dir.join("nextjs-app");

    let rari_result = run_build("rari", &rari_dir, "pnpm run build").await?;
    let nextjs_result = run_build("Next.js", &nextjs_dir, "pnpm run build").await?;

    display_comparison(&rari_result, &nextjs_result);

    let results = BenchmarkResults {
        timestamp: time_utils::format_timestamp(SystemTime::now()),
        rari: rari_result,
        nextjs: nextjs_result,
    };

    save_results(&results, &args.dir).await?;

    println!("\n{}", "🎉 Build comparison completed!".green().bold());

    Ok(())
}
